devinfo/src/partition/mod.rs

```diff
@@ -0,0 +1,330 @@
+    }
+}
+
+impl From<PartitionSource> for &'static str {
+    fn from(pid: PartitionSource) -> &'static str {
+        match pid {
+            PartitionSource::ID => "ID",
+            PartitionSource::Label => "LABEL",
+            PartitionSource::PartLabel => "PARTLABEL",
+            PartitionSource::PartUUID => "PARTUUID",
+            PartitionSource::Path => "PATH",
+            PartitionSource::UUID => "UUID",
+        }
+    }
+}
+
+impl PartitionSource {
+    fn disk_by_path(self) -> PathBuf {
+        PathBuf::from(
+            ["/dev/disk/by-", &<&'static str>::from(self).to_lowercase()]
+                .concat(),
+        )
+    }
+}
+
+/// A collection of all discoverable identifiers for a partition.
+#[derive(Debug, Default, Clone, Hash, PartialEq)]
+pub struct PartitionIdentifiers {
+    pub id: Option<String>,
+    pub label: Option<String>,
+    pub part_label: Option<String>,
+    pub part_uuid: Option<String>,
+    pub path: Option<String>,
+    pub uuid: Option<String>,
+}
+
+impl PartitionIdentifiers {
+    /// Fetches all discoverable identifiers for a partition by the path to that
+    /// partition.
+    pub fn from_path<P: AsRef<Path>>(path: P) -> PartitionIdentifiers {
+        let path = path.as_ref();
+
+        PartitionIdentifiers {
+            path: PartitionID::get_source(SourcePath, path).map(|id| id.id),
+            id: PartitionID::get_source(ID, path).map(|id| id.id),
+            label: PartitionID::get_source(Label, path).map(|id| id.id),
+            part_label: PartitionID::get_source(PartLabel, path)
+                .map(|id| id.id),
+            part_uuid: PartitionID::get_source(PartUUID, path).map(|id| id.id),
+            uuid: PartitionID::get_source(UUID, path).map(|id| id.id),
+        }
+    }
+
+    /// Checks if the given identity matches one of the available identifiers.
+    pub fn matches(&self, id: &PartitionID) -> bool {
+        match id.variant {
+            ID => self.id.as_ref().map_or(false, |s| &id.id == s),
+            Label => self.label.as_ref().map_or(false, |s| &id.id == s),
+            PartLabel => {
+                self.part_label.as_ref().map_or(false, |s| &id.id == s)
+            }
+            PartUUID => self.part_uuid.as_ref().map_or(false, |s| &id.id == s),
+            SourcePath => self.path.as_ref().map_or(false, |s| &id.id == s),
+            UUID => self.uuid.as_ref().map_or(false, |s| &id.id == s),
+        }
+    }
+}
+
+fn attempt<T, F: FnMut() -> Option<T>>(
+    attempts: u8,
+    wait: u64,
+    mut func: F,
+) -> Option<T> {
+    let mut tried = 0;
+    let mut result;
+
+    loop {
+        result = func();
+        if result.is_none() && tried != attempts {
+            ::std::thread::sleep(::std::time::Duration::from_millis(wait));
+            tried += 1;
+        } else {
+            return result;
+        }
+    }
+}
+
+fn canonicalize(path: &Path) -> Cow<'_, Path> {
+    // NOTE: It seems that the kernel may intermittently error.
```

Comment:
Is there a link to an upstream bug report? Same question for the skipping directories issues below.

Reply:
I don't know. This code is mostly imported from the unmaintained partition identity crate. I've only made changes that were needed in order to get it to work for our use case and made sure we only imported stuff we we actually need/use. I'm not even sure if this is stuff we need in the long run as udev can also get all of this info for us.